### PR TITLE
MGMT-3721 ClusterDeployments support update

### PR DESCRIPTION
**Pull request by filanov**

### CODE REVIEW

**Comment by filanov** [comment]
tested with subsystem, i deployed hosts and saw that status was stable. 
```yaml
apiVersion: hive.openshift.io/v1
kind: ClusterDeployment
metadata:
  name: mycluster
  namespace: assisted-installer
spec:
  baseDomain: hive.example.com
  clusterName: mycluster
  installStrategy:
    agent:
      networking:
        clusterNetwork:
          - cidr: 10.128.0.0/14
            hostPrefix: 23
        ServiceNetwork:
          - 172.30.0.0/16
      provisionRequirements:
        controlPlaneAgents: 3
        agentSelector:
          matchLabels:
            bla: aaa
  platform:
    agentBareMetal:
      apiVIP: 1.2.3.8
      apiVIPDNSName: ""
      ingressVIP: 1.2.3.9
      vipDHCPAllocation: false
  provisioning:
    imageSetRef:
      name: openshift-v4.7.0
    installConfigSecretRef:
      name: mycluster-install-config
    sshPrivateKeySecretRef:
      name: mycluster-ssh-key
  pullSecretRef:
    name: pull-secret
```
Result
```yaml
Name:         mycluster
Namespace:    assisted-installer
Labels:       <none>
Annotations:  <none>
API Version:  hive.openshift.io/v1
Kind:         ClusterDeployment
Metadata:
  Creation Timestamp:  2021-02-01T12:42:25Z
  Generation:          1
  Resource Version:    5721
  Self Link:           /apis/hive.openshift.io/v1/namespaces/assisted-installer/clusterdeployments/mycluster
  UID:                 3db39cc8-14c8-418e-9e5d-8496e7f0f295
Spec:
  Base Domain:   hive.example.com
  Cluster Name:  mycluster
  Install Strategy:
    Agent:
      Networking:
        Service Network:
          172.30.0.0/16
        Cluster Network:
          Cidr:         10.128.0.0/14
          Host Prefix:  23
      Provision Requirements:
        Agent Selector:
          Match Labels:
            Bla:               aaa
        Control Plane Agents:  3
  Platform:
    Agent Bare Metal:
      API VIP:              1.2.3.8
      API VIPDNS Name:      bla.com
      Ingress VIP:          1.2.3.9
      Vip DHCP Allocation:  false
  Provisioning:
    Image Set Ref:
      Name:  openshift-v4.7.0
    Install Config Secret Ref:
      Name:  mycluster-install-config
    Ssh Private Key Secret Ref:
      Name:  mycluster-ssh-key
  Pull Secret Ref:
    Name:  pull-secret
Status:
  Conditions:
    Last Probe Time:       2021-02-01T12:48:25Z
    Last Transition Time:  2021-02-01T12:48:25Z
    Message:               ready
    Reason:                AgentPlatformState
    Status:                Unknown
    Type:                  Unreachable
    Last Probe Time:       2021-02-01T12:48:25Z
    Last Transition Time:  2021-02-01T12:48:25Z
    Message:               Cluster ready to be installed
    Reason:                AgnetPlatformStateInfo
    Status:                Unknown
    Type:                  Unreachable
    Last Probe Time:       2021-02-01T12:48:25Z
    Last Transition Time:  2021-02-01T12:48:25Z
    Message:               {"configuration":[{"id":"pull-secret-set","status":"success","message":"The pull secret is set."}],"hosts-data":[{"id":"all-hosts-are-ready-to-install","status":"success","message":"All hosts in the cluster are ready to install."},{"id":"sufficient-masters-count","status":"success","message":"The cluster has a sufficient number of master candidates."}],"network":[{"id":"machine-cidr-defined","status":"success","message":"The Machine Network CIDR is defined."},{"id":"machine-cidr-equals-to-calculated-cidr","status":"success","message":"The Cluster Machine CIDR is equivalent to the calculated CIDR."},{"id":"api-vip-defined","status":"success","message":"The API virtual IP is defined."},{"id":"api-vip-valid","status":"success","message":"api vip 1.2.3.8 belongs to the Machine CIDR and is not in use."},{"id":"ingress-vip-defined","status":"success","message":"The Ingress virtual IP is defined."},{"id":"ingress-vip-valid","status":"success","message":"ingress vip 1.2.3.9 belongs to the Machine CIDR and is not in use."},{"id":"dns-domain-defined","status":"success","message":"The base domain is defined."},{"id":"cluster-cidr-defined","status":"success","message":"The Cluster Network CIDR is defined."},{"id":"service-cidr-defined","status":"success","message":"The Service Network CIDR is defined."},{"id":"no-cidrs-overlapping","status":"success","message":"No CIDRS are overlapping."},{"id":"network-prefix-valid","status":"success","message":"The Cluster Network prefix is valid."},{"id":"ntp-server-configured","status":"success","message":"No ntp problems found"}]}
    Reason:                AgentPlatformCondition
    Status:                Unknown
    Type:                  Unreachable
    Last Probe Time:       2021-02-01T12:47:23Z
    Last Transition Time:  2021-02-01T12:47:23Z
    Message:               failed to update cluster: No suitable matching CIDR found for VIP 1.2.3.8
    Reason:                AgentPlatformError
    Status:                False
    Type:                  Unreachable
  Install Strategy:
    Agent:
      Agent Networks:
        Agent Refs:
          Name:                      165f34c1-fe8c-44b6-8133-6a16e19ccbdc
          Namespace:                 
          Name:                      2d8a0563-ece4-4d11-a458-feb2d5d55d72
          Namespace:                 
          Name:                      357dc5b3-373f-4fa1-b9a3-b63fa1f8eef4
          Namespace:                 
          Name:                      5846cdb8-c30f-4f3f-ac70-70b57faefc9f
          Namespace:                 
          Name:                      b51f769b-2f02-4278-b7a2-e7809278f6f4
          Namespace:                 
        Cidr:                        1.2.3.0/24
      Connectivity Majority Groups:  {"1.2.3.0/24":["165f34c1-fe8c-44b6-8133-6a16e19ccbdc","2d8a0563-ece4-4d11-a458-feb2d5d55d72","357dc5b3-373f-4fa1-b9a3-b63fa1f8eef4","5846cdb8-c30f-4f3f-ac70-70b57faefc9f","b51f769b-2f02-4278-b7a2-e7809278f6f4"]}
Events:                              <none>
```

**Review by nmagnezi — COMMENTED** [review]
Looking good overall.
Few minor comments about tests.

We'll also need to test for failed image updates, but as we discussed it should be a part of `MGMT-3880`

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 1]
```diff
@@ -309,4 +311,135 @@ var _ = Describe("cluster reconcile", func() {
...
+				Status:             corev1.ConditionUnknown,
+				LastProbeTime:      metav1.Time{Time: time.Now()},
+				LastTransitionTime: metav1.Time{Time: time.Now()},
+				Reason:             AgentPlatformState,
+				Message:            models.ClusterStatusPendingForInput,
+			}, &cluster.Status.Conditions)
+			Expect(c.Create(ctx, cluster)).ShouldNot(HaveOccurred())
+		})
+
+		It("update pull-secret network cidr and cluster name", func() {
```
[the remark this answers is not in the document]
Comment (filanov) [reply internal/controller/controllers/clusterdeployments_controller_test.go:341]: Added the validation as part of the mock call.

**Review by nmagnezi — APPROVED** [review]
/lgtm

**Pull request merged**